### ISSUE TITLE
H2o context not created in 3.28.0.1 build with spark 2.4.0-cdh6.2.1

### ISSUE THREAD

**Issue by ananbas** [h2oai/sparkling-water#1763]
Issue: H2o cluster formed but h2ocontext is not created in sparkling-shell

```  
H2O build version         : 3.28.0.1 (yu)
Sparkling Water version   : 3.28.0.1-1-2.4.0-cdh6.2
Spark build version       : 2.4.0-cdh6.2.1
Scala version             : 2.11
Hadoop                    : CDH6.2.1 Hadoop3
Execution mode            : yarn client
OS                        : centos7
```
Spark & Sparkling Water configuration including the memory configuration:
```
spark.blacklist.enabled=false
spark.ext.h2o.cluster.size=10
spark.executor.instances=10
spark.ext.h2o.node.network.mask=10.30.225.0/24
spark.ext.h2o.client.network.mask=10.30.225.0/24
spark.port.maxRetries=50
spark.task.maxFailures=1
spark.scheduler.maxRegisteredResourcesWaitingTime=1000000
spark.sql.autoBroadcastJoinThreshold=-1
spark.locality.wait=0
spark.scheduler.minRegisteredResourcesRatio=1
spark.authenticate=false
spark.dynamicAllocation.enabled=false
spark.eventLog.enabled=true
spark.io.encryption.enabled=false
spark.network.crypto.enabled=false
spark.serializer=org.apache.spark.serializer.KryoSerializer
spark.shuffle.service.enabled=true
spark.shuffle.service.port=7337
spark.ui.enabled=false
spark.ui.killEnabled=false
spark.lineage.log.dir=/var/log/spark/lineage
spark.lineage.enabled=true
spark.default.parallelism=100
spark.driver.memory=2g
spark.driver.memoryOverhead=4g
spark.executor.cores=4
spark.executor.extraJavaOptions=-XX:ParallelGCThreads=4 -XX:+UseParallelGC
spark.executor.memory=30g
spark.executor.memoryOverhead=30g
spark.file.transferTo=false
spark.files.maxPartitionBytes=268435456
spark.network.timeout=600s
spark.rdd.compress=true
spark.rpc.io.serverTreads=64
spark.shuffle.file.buffer=1m
spark.shuffle.io.backLog=8192
spark.shuffle.io.maxRetries=5
spark.shuffle.io.serverThreads=128
spark.shuffle.registration.maxAttempst=5
spark.shuffle.registration.timeout=120000
spark.shuffle.unsafe.file.ouput.buffer=5m
spark.sql.adaptive.enabled=true
spark.sql.parquet.compression.codec=snappy
spark.sql.shuffle.partitions=100
spark.security.credentials.hbase.enabled=false
spark.master=yarn
spark.submit.deployMode=client
```

Code:
```
sc.setLogLevel("INFO")
import org.apache.spark.h2o._
val h2oContext = H2OContext.getOrCreate(spark)
```

Error:
```
java.util.NoSuchElementException: None.get
  at scala.None$.get(Option.scala:347)
  at scala.None$.get(Option.scala:345)
  at org.apache.spark.h2o.SparkSpecificUtils$.addSparklingWaterTab(SparkSpecificUtils.scala:43)
  at org.apache.spark.h2o.H2OContext.init(H2OContext.scala:160)
  at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:606)
  at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:634)
  ... 51 elided
```
[yarn.log](https://github.com/h2oai/sparkling-water/files/4145445/yarn.log)

**Comment by jakubhava** [comment]
Thanks for the report!

This is a simple bug -> we try to register our custom Sparkling Water Tab to the UI even though you have explicitly disabled the UI: `spark.ui.enabled=false` 

As a workaround, you can turn on the UI. We will work on a fix, thanks!

**Comment by ananbas** [comment]
Hi Jakub,

Thank you for replying, after set `spark.ui.enabled=true`, now I get this

```java.lang.RuntimeException: Cloud size 1 under 10
  at water.H2O.waitForCloudSize(H2O.java:1860)
  at org.apache.spark.h2o.backends.internal.InternalH2OBackend$.org$apache$spark$h2o$backends$internal$InternalH2OBackend$$startH2OCluster(InternalH2OBackend.scala:105)
  at org.apache.spark.h2o.backends.internal.InternalH2OBackend.init(InternalH2OBackend.scala:41)
  at org.apache.spark.h2o.H2OContext$H2OContextClientBased.initBackend(H2OContext.scala:448)
  at org.apache.spark.h2o.H2OContext.init(H2OContext.scala:150)
  at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:606)
  at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:634)
  ... 51 elided```

[yarn.log](https://github.com/h2oai/sparkling-water/files/4145482/yarn.log)

**Comment by jakubhava** [comment]
This is error which is going to be fixed in the next release `3.28.0.3`

**Comment by ananbas** [comment]
okay then, thank you Jakub

**Comment by jakubhava** [comment]
no problem, btw: the fix for the UI will be there as well https://github.com/h2oai/sparkling-water/pull/1764

**Comment by ananbas** [comment]
wow, that was fast, thank you Jakub

**Comment by ananbas** [comment]
Hi Jakub,

I'm succefully started h2ocontext with 5 nodes cluster, apparently 10 nodes is too much

**Comment by jakubhava** [comment]
Yup -> the bug which we have fixed behaves like this. In the yarn logs you can see "IndexArrayOutOfBoundExceptions" which appear when more nodes are requested.

This will be fixed in 3.28.0.3 which should go out this week. But if you are not in rush, I would wait for the release 3.28.0.4 which adds additional fixes for clouding in internal backend and making it even more robust

**Comment by jakubhava** [comment]
Sparkling Water 3.28.0.3 is released which fixes the clouding issues mentioned above:

    Spark 2.4: http://h2o-release.s3.amazonaws.com/sparkling-water/spark-2.4/3.28.0.3-1-2.4/index.html
    Spark 2.3: http://h2o-release.s3.amazonaws.com/sparkling-water/spark-2.3/3.28.0.3-1-2.3/index.html
    Spark 2.2: http://h2o-release.s3.amazonaws.com/sparkling-water/spark-2.2/3.28.0.3-1-2.2/index.html
    Spark 2.1: http://h2o-release.s3.amazonaws.com/sparkling-water/spark-2.1/3.28.0.3-1-2.1/index.html

If you bump into any new issues, please create new or feel free to reopen this ticket.
